gqcp/tests/Mathematical/Representation/Tensor_test.cpp

```diff
@@ -281,3 +281,105 @@ BOOST_AUTO_TEST_CASE(addBlock_matrix) {
+    // Start by creating a reference, calculated with numpy.einsum()
+    GQCP::MatrixX<double> reference {2, 2};
+
+    // clang-format off
+    reference <<   11,   35,
+                  131,  155;
+    // clang-format on
+
+    const auto output = T1.einsum<2>(T2, "ijkl", "jk", "il");
+    const auto output2 = T1.einsum<2>("ijkl,jk->il", T2);
+
+
+    BOOST_CHECK(reference.isApprox(output.toMatrix(2, 2), 1.0e-12));
+    BOOST_CHECK(reference.isApprox(output2.toMatrix(2, 2), 1e-12));
+
+    // Test the same for the matrix contraction
+    // Create a matrix containing the same values as T2
+    GQCP::Matrix<double> M {2, 2};
+    for (size_t i = 0; i < 2; i++) {
+        for (size_t j = 0; j < 2; j++) {
+            M(i, j) = j + 5 * i;
+        }
+    }
+
+    const auto output3 = T1.einsum<2>("ijkl,jk->il", M);
+    BOOST_CHECK(reference.isApprox(output3.toMatrix(2, 2), 1e-12));
+
+    // Test the einsum API: single axis contraction, for the same 3 einsum expressions
+    // Start by creating a reference, which was calculated with numpy.einsum()
+    GQCP::Tensor<double, 4> reference_tensor {dim1, dim1, dim1, dim1};
+    for (size_t i = 0; i < dim1; i++) {
+        for (size_t j = 0; j < dim1; j++) {
+            for (size_t k = 0; k < dim1; k++) {
+                for (size_t l = 0; l < dim1; l++) {
+                    reference_tensor(i, j, k, l) = 0;
+                }
+            }
+        }
+    }
+
+    GQCP::MatrixX<double> reference_block_1 {2, 2};
+    // clang-format off
+    reference_block_1 << 50, 60,
+                         60, 74;
+    // clang-format on
+
+    GQCP::MatrixX<double> reference_block_2 {2, 2};
+    // clang-format off
+    reference_block_2 <<   55,   67,
+                           65,   81;
+    // clang-format on
+
+    reference_tensor.addBlock<2, 3>(reference_block_1, 0, 0, 0, 0);
+    reference_tensor.addBlock<2, 3>(reference_block_1, 0, 1, 0, 0);
+    reference_tensor.addBlock<2, 3>(reference_block_2, 1, 0, 0, 0);
+    reference_tensor.addBlock<2, 3>(reference_block_2, 1, 1, 0, 0);
+
+    const auto output4 = T1.einsum<1>(T2, "ijkl", "ia", "jkla");
+    const auto output5 = T1.einsum<1>("ijkl,ia->jkla", T2);
+    const auto output6 = T1.einsum<1>("ijkl,ia->jkla", M);
+
+    BOOST_CHECK(reference_tensor.isApprox(output4, 1e-12));
+    BOOST_CHECK(reference_tensor.isApprox(output5, 1e-12));
+    BOOST_CHECK(reference_tensor.isApprox(output6, 1e-12));
+
+    // Let's test a contraction over all indices as well
+    // We test both einsum notations and the contraction of a tensor with a matrix
+    // reference values were calculated with numpy.einsum()
+    const auto output7 = T1.einsum<4>(T1, "ijkl", "ijkl", "");
+    const auto output8 = T1.einsum<4>("ijkl,ijkl->", T1);
+    const auto output9 = T2.einsum<2>("ij,ij->", M);
+
+    BOOST_CHECK_EQUAL(output7(0), 1096);
+    BOOST_CHECK_EQUAL(output8(0), 1096);
+    BOOST_CHECK_EQUAL(output9(0), 62);
+}
```

Comment:
Check the punctuation and capitalization of the documentation, please.

Also, I see that you're testing multiple things: it's better split this up in different test cases, each specifically testing one part of the API.